### PR TITLE
Integration tests: create new wallet

### PR DESCRIPTION
**Pull request by nkuba**
Depends on https://github.com/keep-network/tbtc-v2/pull/300
Refs https://github.com/keep-network/tbtc-v2/issues/250

In this PR we add an initial integration test to create a new wallet.
The test uses a real implementation of the ECDSA Wallet Registry and mocked Random Beacon.

In another PR we will handle Random Beacon deployment to be able to use the real contract.

Tests for other paths (e.g. slashing) will be added in other PRs.

### CODE REVIEW

File: solidity/test/integration/utils/ecdsa-wallet-registry.ts [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+  await walletRegistry.connect(operator).joinSortitionPool()
+
+  const operatorID = await sortitionPool.getOperatorID(
+    await operator.getAddress()
+  )
+
+  return operatorID
+}
+
+export async function produceEcdsaDkgResult(
```
Comment (pdyraga) [inline solidity/test/integration/utils/ecdsa-wallet-registry.ts:40]: Maybe `performDkg`? It's not only about producing a result that is later submitted by some other function but also about selecting group members, submitting and approving that result. It's a full DKG path.

File: solidity/test/integration/utils/ecdsa-wallet-registry.ts [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+}
+
+async function selectGroup(
+  sortitionPool: Contract,
+  seed: BigNumber
+): Promise<Operator[]> {
+  const identifiers = await sortitionPool.selectGroup(
+    WALLET_SIZE,
+    ethers.utils.hexZeroPad(seed.toHexString(), 32)
+  )
```
Comment (pdyraga) [inline solidity/test/integration/utils/ecdsa-wallet-registry.ts:105]: Why don't we use [`selectGroup` from `WalletRegistry`](https://github.com/keep-network/keep-core/blob/9cf3620dbd8ccd016dc193b2eb23f0a2509dc87c/solidity/ecdsa/contracts/WalletRegistry.sol#L1150-L1152)? This is what the client will call, it does not need to bother about the `SortitionPool`.

Also, if we use this function I think we should be able to get rid of `calculateDkgSeed` and all the complexity calculating and passing the seed introduces.

File: solidity/test/integration/WalleCreation.test.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  before(async () => {
+    ;({ governance, bridge, walletRegistry, randomBeacon } =
+      await waffle.loadFixture(fixture))
+
+    // Update only the parameters that are crucial for this test.
+    await updateWalletRegistryDkgResultChallengePeriodLength(
+      walletRegistry,
+      governance,
+      dkgResultChallengePeriodLength
+    )
```
Comment (pdyraga) [inline solidity/test/integration/WalleCreation.test.ts:51]: And why do we update it in the first place? The test would take too long to execute?

File: solidity/test/integration/WalleCreation.test.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    })
+
+    it("should register a new wallet details in the Bridge", async () => {
+      const storedWallet = await bridge.wallets(walletPubKeyHash)
+
+      expect(storedWallet.ecdsaWalletID).to.be.equal(
+        ecdsaWalletTestData.walletID
+      )
+
+      expect(storedWallet.state).to.be.equal(1)
```
Comment (pdyraga) [inline solidity/test/integration/WalleCreation.test.ts:94]: We can use:

https://github.com/keep-network/tbtc-v2/blob/4cd75f5a691a08de4eb56a16a27ec0f09e1008d4/solidity/test/fixtures/index.ts#L37-L44

Example:
https://github.com/keep-network/tbtc-v2/blob/4cd75f5a691a08de4eb56a16a27ec0f09e1008d4/solidity/test/bridge/Bridge.Wallets.test.ts#L608

File: solidity/test/integration/WalleCreation.test.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+      expect(await bridge.activeWalletPubKeyHash()).to.be.equal(
+        walletPubKeyHash
+      )
+    })
+
+    it("should consume around 93 000 gas for Bridge.requestNewWallet transaction", async () => {
+      await assertGasUsed(requestNewWalletTx, 93_000)
+    })
+
+    // TODO: Should we also validate gas used by Random Beacon and ECDSA Wallet Registry transactions?
```
Comment (pdyraga) [inline solidity/test/integration/WalleCreation.test.ts:117]: At a minimum, the one approving result - it's calling the `Bridge` in a callback.

File: solidity/hardhat.config.ts [hunk 6]
```diff
@@ -55,7 +70,16 @@ const config: HardhatUserConfig = {
...
+      accounts: {
+        // Number of accounts that should be predefined on the testing environment.
+        count:
+          testConfig.nonStakingAccountsCount +
+          testConfig.stakingRolesCount * testConfig.operatorsCount,
+      },
       tags: ["local"],
+      // Ignore contract size on deployment to hardhat network, to be able to
+      // deploy stub contracts in tests.
+      allowUnlimitedContractSize: process.env.TEST_USE_STUBS_TBTC === "true",
```
[the remark this answers is not in the document]
Comment (nkuba) [reply solidity/hardhat.config.ts:82]: Have you also tried running `yarn test`? The unit tests use stubs.

File: solidity/test/integration/utils/ecdsa-wallet-registry.ts (hunk 1 above)
Comment (nkuba) [reply solidity/test/integration/utils/ecdsa-wallet-registry.ts:40]: Good suggestion. I'd name it `performEcdsaDkg` to be more precise, as we will have a similar for the random beacon.

File: solidity/test/integration/utils/ecdsa-wallet-registry.ts (hunk 2 above)
Comment (nkuba) [reply solidity/test/integration/utils/ecdsa-wallet-registry.ts:105]: Nice catch! Yeah, I missed this function. I copied the code from ecdsa's test utils.

File: solidity/test/integration/WalleCreation.test.ts (hunk 3 above)
Comment (nkuba) [reply solidity/test/integration/WalleCreation.test.ts:51]: Yeah exactly, the test hangs for a really long time.

File: solidity/test/integration/utils/ecdsa-wallet-registry.ts (hunk 2 above)
Comment (nkuba) [reply solidity/test/integration/utils/ecdsa-wallet-registry.ts:105]: c8f7d9b60cdd121f41bcac32bf2a2495b668bdad

File: solidity/test/integration/utils/ecdsa-wallet-registry.ts (hunk 1 above)
Comment (pdyraga) [reply solidity/test/integration/utils/ecdsa-wallet-registry.ts:40]: Sure, `performEcdsaDkg` works.

Comment (nkuba) [reply solidity/test/integration/utils/ecdsa-wallet-registry.ts:40]: 006efd4e6d1316b21b322933c16fbff7be5b28e0

File: solidity/test/integration/WalleCreation.test.ts (hunk 4 above)
Comment (nkuba) [reply solidity/test/integration/WalleCreation.test.ts]: e438c884aca08cc30d869c7c7591460771bfa1c9

File: solidity/test/integration/WalleCreation.test.ts (hunk 5 above)
Comment (nkuba) [reply solidity/test/integration/WalleCreation.test.ts]: 04f418bc529cc75f870c133e53dd0b5ac99a9b40

File: solidity/hardhat.config.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/hardhat.config.ts:82]: Unit tests should not be affected by changes in this PR, right?

I have just tried running both `yarn test` and `yarn test:integration` with the two lines earlier mentioned commented out and everything is green.

[the remark this answers is not in the document]
Comment (nkuba) [reply solidity/hardhat.config.ts:82]: You're right! I changed the approach to the interproject deployments so many times that I got confused.

[the remark this answers is not in the document]
Comment (nkuba) [reply solidity/hardhat.config.ts]: 7e5165fef2e9def2af36f7ddbe29c2028a2052a4

[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/hardhat.config.ts]: We should also remove the line at 176:
```
   // Don't check size of contracts used as stubs in tests.
    except: ["contracts/test/.*Stub$"],
```

[the remark this answers is not in the document]
Comment (nkuba) [reply solidity/hardhat.config.ts]: a6cee7a712cad1ba39091fe95f9a3eb36e750317

**Review by pdyraga — APPROVED** [review]

**Pull request merged**